File: data structures/queue/ruby/queue_test.rb
require 'test/unit'
require_relative 'queue'

class QueueTest < Test::Unit::TestCase
  def setup
    @sut = Queue.new
  end

  def test_can_enqueue
    @sut.enqueue 'foo'
    assert_false @sut.empty?
  end

  def test_can_dequeue
    element = 'foo'
    @sut.enqueue element
    assert_equal element, @sut.dequeue
  end

  def test_empty_queue
    assert_true @sut.empty?
    @sut.enqueue 42
    assert_false @sut.empty?
  end

  def test_multiple_inputs
    elements = ['foo', nil, :bar, 42]
    elements.each { |element| @sut.enqueue element }
    assert_false @sut.empty?
    elements.each do |element|
      assert_equal element, @sut.dequeue
    end
    assert_true @sut.empty?
  end
end

File: data structures/queue/python/queue_test.py
import unittest
import queue

class QueueTest(unittest.TestCase):

    def setUp(self):
        self.queue = queue.Queue()     

    def test_enqueue(self):
        self.queue.enqueue(23)
        self.assertFalse(self.queue.isEmpty())

    def test_empty(self):
        self.assertTrue(self.queue.isEmpty())
        self.queue.enqueue(42)
        self.assertFalse(self.queue.isEmpty())
        
    def test_error_when_queue_empty(self):
        with self.assertRaises(AssertionError):
            self.queue.dequeue()

    def test_multiple_enqueue_dequeue(self):
        elements = ['hi', 1, 'hu']

        for element in elements:
            self.queue.enqueue(element)
        
        for element in elements:
            self.assertEqual(self.queue.dequeue(), element)

if __name__ == "__main__":
    unittest.main()